README
NAME
    selfvars - Provide $self, @args, %opts and %hopts variables for OO
    programs

SYNOPSIS
        package MyClass;

        ### Import $self, @args, %opts and %hopts into your package:
        use selfvars;

        ### Or name the variables explicitly:
        # use selfvars -self => 'self', -args => 'args', -opts => 'opts', -hopts => 'hopts';

        ### Write the constructor as usual:
        sub new {
            return bless({}, shift);
        }

        ### Use $self in place of $_[0]:
        sub foo {
            $self->{foo};
        }

        ### Use @args in place of @_[1..$#_]:
        sub bar {
            my ($foo, $bar) = @args;
            $self->{foo} = $foo;
            $self->{bar} = $bar;
        }

        ### Use %opts in place of %{$_[1]}:
        sub baz {
            $self->{x} = $opts{x};
            $self->{y} = $opts{y};
        }
        ### Use %hopts with $obj->yada( x => 1, y => 2 ) call syntax
        sub yada {
            $self->{x} = $hopts{x}
            $self->{y} = $hopts{y}
        }

DESCRIPTION
    This moudles exports three special variables: $self, @args and %opts.

    They are really just handy helpers to get rid of:

        my $self = shift;

    Behind the scenes, $self is simply tied to $_[0], @args to @_[1..$#_],
    and %opts to "%{$_[1]}".

    Currently, both $self and @args are read-only; this means you cannot
    mutate them:

        $self = 'foo';              # error
        my $foo = shift @args;      # error

    This restriction may be lifted at a later version of this module, or
    turned into a configurable option instead.

    However, %opts is not read-only, and can be mutated freely:

        $opts{x} = 'y';             # okay
        delete $opts{x};            # also okay

INTERFACE
    $self
        Returns the current object.

    @args
        Returns the argument list.

    %opts
        Returns the first argument, which must be a hash reference, as a
        hash.

    %hopts
        Returns the arguments list as a hash.

  Choosing non-default names
    You can choose alternative variable names with explicit import
    arguments:

        # Use $this and @vars instead of $self and @args, leaving %opts alone:
        use selfvars -self => 'this', -args => 'vars', -opts;

        # Use $this but leave @args and %opts alone:
        use selfvars -self => 'this', -args, -opts;

        # Use @vars but leave $self and %opts alone:
        use selfvars -args => 'vars', -self, -opts;

    You may also omit a variable name from the explicit import arguments:

        # Import $self but not @args nor %opts:
        use selfvars -self => 'self';

        # Same as the above:
        use selfvars -self;

        # Import $self and %opts but not @args:
        use selfvars -self, -opts;

DEPENDENCIES
    None.

ACKNOWLEDGEMENTS
    This module was inspired and based on Kang-min Liu (gugod)'s "self.pm".

    As seen on #perl:

        <gugod> audreyt: selfvars.pm looks exactly like what I want self.pm to be in the beginning
        <gugod> audreyt: but I can't sort out the last BEGIN{} block like you did.
        <gugod> audreyt: that's a great job :D

SEE ALSO
    self

AUTHORS
    唐鳳 <cpan@audreyt.org>

CC0 1.0 Universal
    To the extent possible under law, 唐鳳 has waived all copyright and
    related or neighboring rights to selfvars.

    This work is published from Taiwan.

    <http://creativecommons.org/publicdomain/zero/1.0>



Changes
[Changes for 0.31 - 2011-05-27]

* Introduce "selfvars::autoload", a helper module that makes it
  possible to write Mojolicious applications with Dancer syntax.

[Changes for 0.22 - 2010-12-14]

* Add =encoding utf8 to POD; no functional changes.

[Changes for 0.21 - 2010-11-29]

* Improved POD for %hopts; no functional changes.

* Trying to mutate %hopts now fails with this error message:
    "Modification of a read-only %hopts attempted"
  which was erroneously listed as "%args" in 0.20.

[Changes for 0.20 - 2010-11-23]

* Added %hopts for ($self, key => value) styled calls:

    package Foo;
    use selfvars;
    sub meth {
        $self->{x} = $hopts{x};
    }

    # Call it like this:
    $foo->meth(x => 4);

  Contributed by Roman Galeev.

[Changes for 0.11 - 2010-07-09]

* LICENSING CHANGE: This compilation and all individual files in it
  are now under the nullary CC0 1.0 Universal terms:

  To the extent possible under law, 唐鳳 has waived all copyright and
  related or neighboring rights to selfvars.

* Documentation fix: %opt stands for %{$_[1]}, not %{$_[0]}. (Luke Closs)

* Update minimum Perl version from 5.004 to 5.005 as required by Module::Install.

[Changes for 0.10 - 2007-12-29]

* Add support for %opts, an alias for named arguments in %{$_[1]}:

    package Foo;
    use selfvars;
    sub meth {
        $self->{x} = $opts{x};
    }

    # Call it like this:
    $foo->meth({x => 4});

* Support for omitting the "undef" part in selective imports:

    # Import $self and %opts but not @args
    use selfvars -self, -opts;

[Changes for 0.06 - 2007-12-29]

* Minor test suite fixes for Perl 5.004/5.005 compatibility.

[Changes for 0.05 - 2007-12-28]

* You can now choose alternate names for $self and @args:

    use selfvars -self => 'this', -args => 'vars';

[Changes for 0.04 - 2007-12-28]

* Further fixes for supporting Perl 5.005 and 5.004.

[Changes for 0.03 - 2007-12-28]

* Documented that $self and @args are read-only variables for now.

* @args is made truly read-only.

[Changes for 0.02 - 2007-12-28]

* Fix support for Perl 5.005 and 5.004.

[Changes for 0.01 - 2007-12-27]

* Initial release to CPAN of this six-minutes-old hack.


t/lib/HOpts.pm
use strict;

package HOpts;
use selfvars -self => 'this', -args => 'vars', -opts => 'vars', -hopts => 'args';

sub new {
    my $class = shift;
    return bless { v => 0 }, $class;
}

sub hopts {
	return join ' ', $this->{v}, $args{'param'};
}

1;


lib/selfvars/autoload.pm
package selfvars::autoload;
use 5.005;
use strict;
use selfvars ();

sub import {
    no strict 'refs';
    my $pkg = caller;
    *{"$pkg\::AUTOLOAD"} = \&autoload;
    shift; unshift @_, 'selfvars';
    goto &selfvars::import;
}

sub autoload {
    no strict 'vars';
    my $method = $AUTOLOAD;

    my $self = do {
        package DB;
        my $i = 1;
        () = caller($i);
        $DB::args[0];
    } or do {
        require Carp;
        Carp::croak("Undefined subroutine &$AUTOLOAD called");
    };

    $method =~ s/.*:://;
    if (my $code = $self->can($method)) {
        unshift @_, $self;
        goto &$code;
    }
    $self->$method(@_); # ...and let it fail...
}

1;

__END__

=encoding utf8

=head1 NAME

selfvars::autoload - Turn missing_sub(...) into $_[0]->missing_sub(...) implicitly

=head1 SYNOPSIS

    use Mojolicious::Lite; # The raison d'être for this module...

    # Import $self, @args, %opts and %hopts into your package;
    # see "perldoc selfvars" for import options and usage.
    use selfvars::autoload;

    # Normal invocation with two "$self"s
    get '/' => sub {
        my $self = shift;
        $self->render(text => 'Hello World!');
    };

    # It's OK to omit "my $self = shift":
    get '/selfish' => sub {
        $self->render(text => 'Hello World!');
    };

    # It's OK to omit the "$self->" part too!
    get '/selfless' => sub {
        render(text => 'Hello World!');
    };

    # dance!
    app->start;

=head1 DESCRIPTION

This module exports four special variables: C<$self>, C<@args>, C<%opts> and C<%hopts>;
see L<selfvars> for the full description and import options.

In addition to that, this module sets up an C<AUTOLOAD> subroutine in the importing
package, so any calls to missing functions becomes a method call with C<$_[0]> as
the invocant.

If C<$_[0]> is not present, then we raise an C<Undefined subroutine> exception as usual.

The net effect is that we can start writing Mojolicious apps with Dancer syntax. :-)

=head1 DEPENDENCIES

None.

=head1 SEE ALSO

L<selfvars>, L<Mojolicious>, L<Dancer>

=head1 AUTHORS

唐鳳 E<lt>cpan@audreyt.orgE<gt>

=head1 CC0 1.0 Universal

To the extent possible under law, 唐鳳 has waived all copyright and related
or neighboring rights to selfvars.

This work is published from Taiwan.

L<http://creativecommons.org/publicdomain/zero/1.0>

=cut


lib/selfvars.pm
package selfvars;
use 5.005;
use strict;
use vars qw( $VERSION $self @args %opts %hopts );

BEGIN {
    $VERSION = '0.31';
}

sub import {
    my $class = shift; # Oooh, the irony!
    my %vars  = (-self => undef, -args => undef, -opts => undef, -hopts => undef) unless @_;

    while (@_) {
        my $key = shift;
        if (@_ and $_[0] !~ /^-/) {
            $vars{$key} = shift;
        }
        else {
            $vars{$key} = undef;
        }
    }

    my $pkg = caller;

    no strict 'refs';
    my %map = (self => \$self, args => \@args, opts => \%opts, hopts => \%hopts);
    while (my ($sym, $var) = each %map) {
        exists $vars{"-$sym"} or next;
        $vars{"-$sym"} = $sym unless defined $vars{"-$sym"};
        *{"$pkg\::$vars{qq[-$sym]}"} = $var;
    }
}

package selfvars::self;

sub TIESCALAR {
    my $x;
    bless \$x => $_[0];
}

sub FETCH {
    my $level = 1;
    my @c     = ();
    while ( !defined( $c[3] ) || $c[3] eq '(eval)' ) {
        @c = do {
            package DB;
            @DB::args = ();
            caller($level);
        };
        $level++;
    }
    $DB::args[0];
}

sub STORE {
    require Carp;
    Carp::croak('Modification of a read-only $self attempted');
}

package selfvars::args;
use Tie::Array ();
use vars qw(@ISA);
BEGIN { @ISA = 'Tie::Array' }

sub _args {
    my $level = 2;
    my @c;
    while ( !defined( $c[3] ) || $c[3] eq '(eval)' ) {
        @c = do {
            package DB;
            @DB::args = ();
            caller($level);
        };
        $level++;
    }
    \@DB::args;
}

sub readonly { require Carp; Carp::croak('Modification of a read-only @args attempted'); }

sub TIEARRAY  { my $x; bless \$x => $_[0] }
sub FETCHSIZE { scalar $#{ _args() } }
sub STORESIZE { goto &readonly } # $#{ _args() } = $_[1] + 1;
sub STORE     { _args()->[ $_[1] + 1 ] = $_[2] }
sub FETCH     { _args()->[ $_[1] + 1 ] }
sub CLEAR     { goto &readonly } # $#{ _args() } = 0; 
sub POP       { goto &readonly } # my $o = _args(); (@$o > 1) ? pop(@$o) : undef
sub PUSH      { goto &readonly } # my $o = _args(); push( @$o, @_ )
sub SHIFT     { goto &readonly } # my $o = _args(); splice( @$o, 1, 1 ) 
sub UNSHIFT   { goto &readonly } # my $o = _args(); unshift( @$o, @_ ) 
sub DELETE    { goto &readonly } # my $o = _args(); delete $o->[ $_[1] + 1 ]
sub SPLICE    { goto &readonly } 
    # my $ob  = shift;
    # my $sz  = $ob->FETCHSIZE;
    # my $off = @_ ? shift : 0;
    # $off += $sz if $off < 0;
    # my $len = @_ ? shift : $sz - $off;
    # splice( @$ob, $off + 1, $len, @_ );

BEGIN {
    local $@;
    eval q{
        sub EXISTS    {
            my $o = _args(); exists $o->[ $_[1] + 1 ]
        }
    } if $] >= 5.006;
}

package selfvars::opts;

sub _opts {
    my $level = 2;
    my @c;
    while ( !defined( $c[3] ) || $c[3] eq '(eval)' ) {
        @c = do {
            package DB;
            @DB::args = ();
            caller($level);
        };
        $level++;
    }
    $DB::args[1];
}

sub TIEHASH  { my $x; bless \$x => $_[0] }
sub FETCH    { _opts()->{ $_[1] } }
sub STORE    { _opts()->{ $_[1] } = $_[2] }
sub FIRSTKEY { my $o = _opts(); my $a = scalar keys %$o; each %$o }
sub NEXTKEY  { my $o = _opts(); each %$o }
sub EXISTS   { my $o = _opts(); exists $o->{$_[1]} }
sub DELETE   { my $o = _opts(); delete $o->{$_[1]} }
sub CLEAR    { my $o = _opts(); %$o = () }
sub SCALAR   { my $o = _opts(); scalar %$o }

package selfvars::hopts;

sub _opts {
    my $level = 2;
    my @c;
    while ( !defined( $c[3] ) || $c[3] eq '(eval)' ) {
        @c = do {
            package DB;
            @DB::args = ();
            caller($level);
        };
        $level++;
    }
    shift @DB::args;
    @DB::args;
}

sub readonly { require Carp; Carp::croak('Modification of a read-only %hopts attempted'); }

sub TIEHASH  { my $x; bless \$x => $_[0] }
sub FETCH    { my (%o) = _opts(); $o{ $_[1] } }
sub STORE    { goto &readonly }
sub FIRSTKEY { my (%o) = _opts(); my $a = scalar keys %o; each %o }
sub NEXTKEY  { }
sub EXISTS   { my (%o) = _opts(); exists $o{$_[1]} }
sub DELETE   { goto &readonly }
sub CLEAR    { goto &readonly }
sub SCALAR   { my (%o) = _opts(); scalar %o }


package selfvars;

BEGIN {
    tie $self => __PACKAGE__ . '::self';
    tie @args => __PACKAGE__ . '::args';
    tie %opts => __PACKAGE__ . '::opts';
    tie %hopts => __PACKAGE__ . '::hopts';
}

1;

__END__

=encoding utf8

=head1 NAME

selfvars - Provide $self, @args, %opts and %hopts variables for OO programs

=head1 SYNOPSIS

    package MyClass;

    ### Import $self, @args, %opts and %hopts into your package:
    use selfvars;

    ### Or name the variables explicitly:
    # use selfvars -self => 'self', -args => 'args', -opts => 'opts', -hopts => 'hopts';

    ### Write the constructor as usual:
    sub new {
        return bless({}, shift);
    }

    ### Use $self in place of $_[0]:
    sub foo {
        $self->{foo};
    }

    ### Use @args in place of @_[1..$#_]:
    sub bar {
        my ($foo, $bar) = @args;
        $self->{foo} = $foo;
        $self->{bar} = $bar;
    }

    ### Use %opts in place of %{$_[1]}:
    sub baz {
        $self->{x} = $opts{x};
        $self->{y} = $opts{y};
    }

    ### Use %hopts with $obj->yada( x => 1, y => 2 ) call syntax
    sub yada {
        $self->{x} = $hopts{x}
        $self->{y} = $hopts{y}
    }

=head1 DESCRIPTION

This module exports four special variables: C<$self>, C<@args>, C<%opts> and C<%hopts>.

They are really just handy helpers to get rid of:

    my $self = shift;

Behind the scenes, C<$self> is simply tied to C<$_[0]>, C<@args> to
C<@_[1..$#_]>, C<%opts> to C<%{$_[1]}>, and C<%hopts%> to C<%{{@_[1..$#_]}}>.

Currently, C<$self>, C<@args> and C<%hopts> are read-only; this means you cannot
mutate them:

    $self = 'foo';              # error
    my $foo = shift @args;      # error
    $hopts{x} = 'y';            # error
    delete $hopts{x};           # error

This restriction may be lifted at a later version of this module, or turned
into a configurable option instead.

However, C<%opts> is not read-only, and can be mutated freely:

    $opts{x} = 'y';             # okay
    delete $opts{x};            # also okay

=head1 INTERFACE

=over 4

=item $self

Returns the current object.

=item @args

Returns the argument list.

=item %opts

Returns the first argument, which must be a hash reference, as a hash.

=item %hopts

Returns the arguments list as a hash.

=back

=head2 Choosing non-default names 

You can choose alternative variable names with explicit import arguments:

    # Use $this and @vars instead of $self and @args, leaving %opts and %hopts alone:
    use selfvars -self => 'this', -args => 'vars', -opts, -hopts;

    # Use $this but leave @args, %opts and %hopts alone:
    use selfvars -self => 'this', -args, -opts, -hopts;

    # Use @vars but leave $self, %opts and %hopts alone:
    use selfvars -args => 'vars', -self, -opts, -hopts;

You may also omit one or more variable names from the explicit import arguments:

    # Import $self but not @args, %opts nor %hopts:
    use selfvars -self => 'self';

    # Same as the above:
    use selfvars -self;

    # Import $self and %opts but not @args nor %hopts:
    use selfvars -self, -opts;

=head1 DEPENDENCIES

None.

=head1 ACKNOWLEDGEMENTS 

This module was inspired and based on Kang-min Liu (gugod)'s C<self.pm>.

As seen on #perl:

    <gugod> audreyt: selfvars.pm looks exactly like what I want self.pm to be in the beginning
    <gugod> audreyt: but I can't sort out the last BEGIN{} block like you did.
    <gugod> audreyt: that's a great job :D

=head1 SEE ALSO

L<self>

=head1 AUTHORS

唐鳳 E<lt>cpan@audreyt.orgE<gt>

=head1 CC0 1.0 Universal

To the extent possible under law, 唐鳳 has waived all copyright and related
or neighboring rights to selfvars.

This work is published from Taiwan.

L<http://creativecommons.org/publicdomain/zero/1.0>

=cut
